crates/typst/src/geom/color.rs

```diff
@@ -334,183 +954,696 @@ cast! {
-        Self { r, g, b, a }
+    fn saturate(self, factor: Ratio) -> Self {
+        self.to_hsv().saturate(factor).to_rgba()
     }
 
-    /// Lighten this color by a factor.
-    ///
-    /// The alpha channel is not affected.
-    pub fn lighten(self, factor: Ratio) -> Self {
-        let lighten =
-            |c: u8| c.saturating_add(round_u8((u8::MAX - c) as f64 * factor.get()));
-        Self {
-            r: lighten(self.r),
-            g: lighten(self.g),
-            b: lighten(self.b),
-            a: self.a,
-        }
+    fn desaturate(self, factor: Ratio) -> Self {
+        self.to_hsv().desaturate(factor).to_rgba()
     }
 
-    /// Darken this color by a factor.
-    ///
-    /// The alpha channel is not affected.
-    pub fn darken(self, factor: Ratio) -> Self {
-        let darken = |c: u8| c.saturating_sub(round_u8(c as f64 * factor.get()));
-        Self {
-            r: darken(self.r),
-            g: darken(self.g),
-            b: darken(self.b),
-            a: self.a,
-        }
+    fn hue_rotate(self, hue: Angle) -> Self {
+        self.to_hsv().hue_rotate(hue).to_rgba()
     }
 
-    /// Negate this color.
-    ///
-    /// The alpha channel is not affected.
-    pub fn negate(self) -> Self {
-        Self {
-            r: u8::MAX - self.r,
-            g: u8::MAX - self.g,
-            b: u8::MAX - self.b,
-            a: self.a,
+    fn negate(self) -> Self {
+        match self {
+            Self::Floating { r, g, b, a } => Self::new(1.0 - r, 1.0 - g, 1.0 - b, a),
+            Self::Integer { r, g, b, a } => {
+                Self::Integer { r: u8::MAX - r, g: u8::MAX - g, b: u8::MAX - b, a }
+            }
         }
     }
 
-    /// Converts this color to a RGB Hex Code.
-    pub fn to_hex(self) -> EcoString {
-        if self.a != 255 {
-            eco_format!("#{:02x}{:02x}{:02x}{:02x}", self.r, self.g, self.b, self.a)
+    fn to_hex(self) -> EcoString {
+        let (r, g, b, a) = match self {
+            Self::Floating { r, g, b, a } => (
+                round_u8(r * 255.0),
+                round_u8(g * 255.0),
+                round_u8(b * 255.0),
+                round_u8(a * 255.0),
+            ),
+            Self::Integer { r, g, b, a } => (r, g, b, a),
+        };
+        if self.alpha() != Some(1.0) {
+            eco_format!("#{:02x}{:02x}{:02x}{:02x}", r, g, b, a)
         } else {
-            eco_format!("#{:02x}{:02x}{:02x}", self.r, self.g, self.b)
+            eco_format!("#{:02x}{:02x}{:02x}", r, g, b)
         }
     }
+}
+
+impl RgbaColor {
+    /// Construct a new RGBA color.
+    pub fn new(r: f64, g: f64, b: f64, a: f64) -> Self {
+        Self::Floating { r, g, b, a }
+    }
 
-    /// Converts this color to an array of R, G, B, A components.
-    pub fn to_array(self) -> Array {
-        array![self.r, self.g, self.b, self.a]
+    /// Construct a new RGBA color from 8-bit values.
+    pub const fn new_from_u8(r: u8, g: u8, b: u8, a: u8) -> Self {
```

Comment:
I generally dislike function names that contain `new` _and_ `from`. I would remove the `new_` prefix. That aside, just it's actually multiple u8s, but `from_u8s` doesn't read that nicely ...